cirq-core/cirq/ops/controlled_gate.py

```diff
@@ -143,7 +146,9 @@ def _decompose_(self, qubits):
         ):
             control_qubits = list(qubits[: self.num_controls()])
             invert_ops: List['cirq.Operation'] = []
-            for cvals, cqbit in zip(self.control_values, qubits[: self.num_controls()]):
+            for cvals, cqbit in zip(
+                self.control_values._identifier(), qubits[: self.num_controls()]
```

Comment:
Iterating on `self.control_values._identifier()` assumes that the internal representation is factorized into qubits. This is only true for the `ProductOfSums` derived class. We should have an isinstance check in the if condition above before relying on this implementation specific assumption. 


Same for the usage of `._identifier()` below.

Reply:
done